txpipe/source_selector.py

```diff
@@ -673,9 +469,252 @@ def select_2d(self, data, calling_from_select=False):
+
+        return pz_data
+
+    def compute_output_stats(self, calculator, mean, variance):
+        R, S, N = calculator.collect(self.comm)
+        calibrator = MetaCalibrator(R, S, mean, mu_is_calibrated=False)
+        mean_e = calibrator.mu
+
+        Rtot = R + S
+        P = np.diag(np.linalg.inv(Rtot @ Rtot))
+        # Apply to the variances to get sigma_e
+        sigma_e = np.sqrt(0.5 * P @ variance)
+
+        return BinStats(N, N, mean_e, sigma_e, calibrator)
+
+
+
+
+class TXSourceSelectorMetadetect(TXSourceSelectorBase):
+    name = "TXSourceSelectorMetadetect"
+    def data_iterator(self):
+        chunk_rows = self.config['chunk_rows']
+        bands = self.config['bands']
+        shear_cols = metadetect_variants(
+                'T', 's2n', 'g1', 'g2',
+                'ra', 'dec', 'mcal_psf_T_mean', 'weight', 'flags'
+                )
+        shear_cols += band_variants(bands, 'mag', 'mag_err', shear_catalog_type='metadetect')
+
+        if self.config['input_pz']:
+            shear_cols += metadetect_variants('mean_z')
+        elif self.config['true_z']:
+            shear_cols += metadetect_variants('redshift_true')
+
+        renames = {}
+        for prefix in ['00', '1p', '1m', '2p', '2m']:
+            renames[f'{prefix}/mcal_psf_T_mean'] = f'{prefix}/psf_T_mean'
+
+        it = self.iterate_hdf('shear_catalog', 'shear', shear_cols, chunk_rows, longest=True)
+        return rename_iterated(it, renames)
+
+    def setup_response_calculators(self, nbin_source):
+        delta_gamma = self.config['delta_gamma']
+        calculators = [MetaDetectCalculator(self.select, delta_gamma) for i in range(nbin_source)]
+        calculators.append(MetaDetectCalculator(self.select_2d, delta_gamma))
+        return calculators
+
+    def apply_simple_redshift_cut(self, data):
+        # If we have the truth pz then we just need to do the binning once,
+        # as in the parent class
+        if not self.config['input_pz']:
+            return super().apply_simple_redshift_cut(data)
+
+        # Otherwise we have to do it once for each variant
+        pz_data = {}
+        variants = ['00/', '1p/', '2p/', '1m/', '2m/']
+        for v in variants:
+            zz = shear_data[f'{v}mean_z']
+
+            pz_data_v = np.zeros(len(zz), dtype=int) -1
+            for zi in range(len(self.config['source_zbin_edges'])-1):
+                mask_zbin = (zz>=self.config['source_zbin_edges'][zi]) & (zz<self.config['source_zbin_edges'][zi+1])
+                pz_data_v[mask_zbin] = zi
+
+            pz_data[f'{v}zbin'] = pz_data_v
+
+        return pz_data
+
+
+    def setup_output(self):
+        outfile = super().setup_output()
+        n = outfile['tomography/source_bin'].size
+        nbin_source = outfile['tomography/source_counts'].size
+        group = outfile.create_group('response')
+        group.create_dataset('R', (nbin_source,2,2), dtype='f')
+        group.create_dataset('R_2d', (2,2), dtype='f')
```

Comment:
do we not need the other quantities in the metacal output?

Reply:
Yep - these are done in the superclass call at the top. I'll adda comment.